server/Controllers/GameControllers/snake.js

```diff
@@ -1,13 +1,126 @@
+   * @param  {number} initialPositionX
+   * @param  {number} initialPositionY
+   * @param  {string} initialDirection
+   * @param  {string} snakeColor
+   * @param  {number} initialSnakeLength
+   * @param  {Function} eatFood Function to check that snake's next movement will be eating food.
+   * @param  {Function} hasCollidedWithRival Function to check that snake's next movement is colliding with Rival's body or not
+   * @param  {number} intervalNumber
+   */
   constructor(playerId, initialPositionX, initialPositionY,
-    initialDirection, snakeColor, initialSnakeLength) {
+    initialDirection, snakeColor, initialSnakeLength, eatFood, hasCollidedWithRival, intervalNumber) {
     this.ownerId = playerId;
     this.head = { x: initialPositionX, y: initialPositionY };
     this.direction = initialDirection;
     this.color = snakeColor;
     this.bodyLength = initialSnakeLength;
+    this.eatFood = eatFood;
+    this.hasCollidedWithRival = hasCollidedWithRival;
+    this.interval = intervalNumber; // interval variable, if snake collides we will clear this interval
+    this.xAxisVelocity = 0; // the x axis movement of snake in each interval
+    this.yAxisVelocity = 0; // the y axis movement of snake in each interval
     this.bodyCoordinates = [];
   }
+
+  /**
+   * This function generates snake body coordinates from the given starting head coordinates
+   */
+  createInitialSnakeBody() {
+    switch (this.direction) {
+      case 'down': {
+        this.yAxisVelocity = 2;
+        break;
+      }
+      case 'up': {
+        this.yAxisVelocity = -2;
+        break;
+      }
+      case 'left': {
+        this.xAxisVelocity = -2;
+        break;
+      }
+      case 'right': {
+        this.xAxisVelocity = 2;
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+    // FirstRender of snake body
+    for (let x = 0; x < this.length; x += 1) {
+      this.moveSnakeOneStep();
+    }
+  }
+
+  /**
+   * Moves the snakes one steps according to relative change on x axis or y axis.
+   */
+  moveSnakeOneStep() {
+    this.head.x += this.xAxisVelocity; // snake head's new x coordinate
+    this.head.y += this.yAxisVelocity; // snake head's new y coordinate
+    if (this.eatFood(this.head)) {
+      this.length += 15;
+      // @TODO implement the score/lifeline increment here.
+    }
+    if (this.hasCollidedWithItself()
+      || this.hasCollidedWithRival(this.head)) {
+      // @TODO emit a game mechanics event to show the modal for game lost.
+
+    } else {
+      this.pushCoordinates();
+    }
+  }
+
+  /**
+   * Pushing new coordinates to body coordinates array of snake
+   */
+  pushCoordinates() {
+    this.bodyCoordinates.push({
+      x: this.head.x,
+      y: this.head.y,
+    });
+    this.snakeLengthControl();
+  }
+
+  /**
+   * strips etc length from snake's body that is more than current length of snake
+   */
+  snakeLengthControl() {
+    if (this.bodyCoordinates.length > this.length) {
```

Comment:
should this be a`while` instead of `if`?

Reply:
No, it is just a check to see if the current body length is actually greater than snake length than only we will strip the snake body. and if happens on every step.

Reply:
Yeah, but since you're calling `this.bodyCoordinates.shift();` meaning you're removing one unit of snake body. It will be shortened by only one unit. 

But what if the current body is two units longer than expected body length? It will only be shortened by one unit and will not match. Does this make sense?

Reply:
The body can't go two unites longer than expected body length, as soon as it goes one length up from a length, the array shifts one position.

Reply:
It just stripes the old coordinate from the snake body. For example, if the body length is 30, then there are 30  coordinates in the body. And when the snake moves(pushCoordinate function), It adds one coordinate in the body, but the snake length has to be 30. So I am just striping the oldest coordinate(tail coordinate) from the snake body array.